apps/rush-lib/src/logic/DependencyIntegrator.ts

```diff
@@ -0,0 +1,355 @@
+    } else {
+      console.log(`No version selector specified, will be automatically determined.`);
+    }
+    console.log();
+
+    // if ensureConsistentVersions => reuse the pinned version
+    // else, query the registry and use the latest that satisfies semver spec
+    if (initialSpec && implicitlyPinnedVersion && initialSpec === implicitlyPinnedVersion) {
+      console.log(colors.green('The specified version ')
+        + colors.magenta(initialSpec)
+        + colors.green(' has been selected as it matches the implicitly preferred version.'));
+      return initialSpec;
+    }
+
+    if (this._rushConfiguration.ensureConsistentVersions && !initialSpec && implicitlyPinnedVersion) {
+      console.log(colors.grey('The enforceConsistentVersions policy is currently active.'));
+      console.log(`Using the implicitly preferred version ${colors.magenta(implicitlyPinnedVersion)}`);
+      return implicitlyPinnedVersion;
+    }
+
+    let selectedVersion: string | undefined;
+
+    if (initialSpec && initialSpec !== 'latest') {
+      console.log(colors.gray('Finding newest version that satisfies the selector: ') + initialSpec);
+      console.log();
+      console.log(`Querying registry for all versions of ${packageName}...`);
+      const allVersions: string =
+        Utilities.executeCommandAndCaptureOutput(this._rushConfiguration.packageManagerToolFilename,
+          ['view', packageName, 'versions', '--json'],
+          this._rushConfiguration.commonTempFolder);
+
+      let versionList: Array<string> = JSON.parse(allVersions);
+      versionList = versionList.sort((a: string, b: string) => { return semver.gt(a, b) ? -1 : 1; });
+
+      console.log(colors.gray(`Found ${versionList.length} available versions.`));
+
+      for (const version of versionList) {
+        if (semver.satisfies(version, initialSpec)) {
+          selectedVersion = version;
+          console.log(`Found latest version: ${colors.magenta(selectedVersion)}`);
+          break;
+        }
+      }
+      if (!selectedVersion) {
+        throw new Error(`Cannot find version for ${packageName} that satisfies '${initialSpec}'`);
+      }
+    } else {
+      if (initialSpec !== 'latest') {
+        console.log(colors.gray(`The enforceConsistentVersions policy is NOT active,`
+          + ` therefore using the latest version.`));
+        console.log();
+      }
+      console.log(`Querying NPM registry for latest version of ${packageName}...`);
+
+      selectedVersion = Utilities.executeCommandAndCaptureOutput(this._rushConfiguration.packageManagerToolFilename,
+        ['view', `${packageName}@latest`, 'version'],
+        this._rushConfiguration.commonTempFolder).trim();
+      console.log();
+
+      console.log(`Found latest version: ${colors.magenta(selectedVersion)}`);
+    }
+
+    console.log();
+
+    if (rangeStyle === SemVerStyle.Caret) {
+      console.log(colors.gray('The --caret flag was specified, prepending ^ specifier to version.'));
```

Comment:
>-care [](start = 36, length = 5)

"--caret"

Reply:
Generally we try to use `${this._caretFlag.longName}`  in case the parameter gets renamed later.  Is there some way to access that here?  (Don't bother if this is a disruptive change.)

---
In reply to: [220771026](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220771026) [](ancestors = 220771026)

Reply:
Not super easy to access this in a different class without piping this around

---
In reply to: [220771228](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220771228) [](ancestors = 220771228,220771026)